tools/clang/lib/SPIRV/DeclResultIdMapper.cpp

```diff
@@ -1900,6 +1905,24 @@ bool DeclResultIdMapper::createStageVars(
               astContext.getExtVectorType(srcVecElemType, 2), *value, *value,
               {0, 1}, thisSemantic.loc);
       }
+      // Special handling of SV_ShadingRate, which is a bitpacked enum value,
+      // but SPIR-V's FragSizeEXT uses a uint2. We build the enum value from
+      // this enum.
+      else if (semanticKind == hlsl::Semantic::Kind::ShadingRate) {
+        // From the D3D12 functional spec for Variable-Rate Shading.
+        // #define D3D12_MAKE_COARSE_SHADING_RATE(x,y) ((x) << 2 | (y))
+        const auto x = spvBuilder.createCompositeExtract(
+            astContext.IntTy, *value, {0}, thisSemantic.loc);
+        const auto y = spvBuilder.createCompositeExtract(
+            astContext.IntTy, *value, {1}, thisSemantic.loc);
+        const auto constTwo =
+            spvBuilder.getConstantInt(astContext.IntTy, llvm::APInt(32, 2));
+        *value = spvBuilder.createBinaryOp(
+            spv::Op::OpBitwiseOr, astContext.UnsignedIntTy,
```

Comment:
All the types used here seem to be signed int type -- except for the result of the BitwiseOr which is unsigned int. Is this intentional? Should all of them be uint?

Reply:
The original specification simply says "two-component vector of 32-bit integers" [0], so I took it to mean signed. HLSL's enums, I believe, are unsigned. I didn't quite know where to put the conversion, but this appeared to work.

https://renderdoc.org/vkspec_chunked/chap14.html#interfaces-builtin-variables-fragsize